### PR TITLE
sched：remove g_cpu_schedlock g_cpu_irqsetlock g_cpu_locksetlock

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
we can use g_cpu_lockset to determine whether we are currently in the scheduling lock, and all accesses and modifications to g_cpu_lockset, g_cpu_irqlock, g_cpu_irqset are in the critical section, so we can directly operate on it.

## Testing
test:
We can use qemu for testing.

compiling
make distclean -j20; ./tools/configure.sh -l qemu-armv8a:nsh_smp ;make -j20 running
qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel ./nuttx

### CODE REVIEW

File: sched/irq/irq_csection.c [hunk 1]
```diff
@@ -615,13 +613,14 @@ void restore_critical_section(void)
...
 
-  FAR struct tcb_s *tcb = this_task();
+  FAR struct tcb_s *tcb;
   int me = this_cpu();
 
   /* Adjust global IRQ controls.  If irqcount is greater than zero,
    * then this task/this CPU holds the IRQ lock
    */
 
+  tcb = current_task(me);
```
Comment (masayuki2009) [inline sched/irq/irq_csection.c:623]: @hujun260 
Please tell us why did you change this line in this PR.

Comment (hujun260) [reply sched/irq/irq_csection.c:623]: this_task() is a function call that involves disabling interrupts, and since restore_critical_section is always called in an interrupt-disabled context, there's no need to disable interrupts again.

Therefore, to save time and achieve the same effect, I directly call tcb = current_task(me) instead of tcb = this_task().

Comment (xiaoxiang781216) [reply sched/irq/irq_csection.c:623]: @masayuki2009 do you have more concern

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

File: sched/irq/irq_csection.c (hunk 1 above)
Comment (masayuki2009) [reply sched/irq/irq_csection.c:623]: @hujun260 
Please add a separate commit (not a PR) for it.

Comment (hujun260) [reply sched/irq/irq_csection.c:623]: > @hujun260 Please add a separate commit (not a PR) for it.

ok

Comment (masayuki2009) [reply sched/irq/irq_csection.c:623]: @hujun260 
Thanks for dividing the commit.
However, I think the new commit message is still unclear and has long lines.

So, could you change the 2nd commit message as follows?

```
Author: hujun5 <hujun5@xiaomi.com>
Date:   Tue Jun 11 08:05:26 2024 +0800

    sched: irq: Change this_task to current_task(me) in restore_critical_section()
    
    this_task() is a function call that involves disabling interrupts and this_cpu().
    Since restore_critical_section is always called in an interrupt-disabled context,
    there's no need to disable interrupts again. Therefore, to save time and achieve
    the same effect, I directly call tcb = current_task(me) instead of tcb = this_task().
    
    Signed-off-by: hujun5 <hujun5@xiaomi.com>
```

Comment (hujun260) [reply sched/irq/irq_csection.c:623]: ok

File: sched/sched/sched_lock.c [hunk 2]
```diff
@@ -171,18 +131,17 @@ int sched_lock(void)
...
           /* If this thread already has the scheduler locked, then
-           * g_cpu_schedlock() should indicate that the scheduler is locked
+           * g_cpu_lockset should indicate that the scheduler is locked
            * and g_cpu_lockset should include the bit setting for this CPU.
            */
 
-          DEBUGASSERT(spin_is_locked(&g_cpu_schedlock) &&
-                      (g_cpu_lockset & (1 << this_cpu())) != 0);
+          DEBUGASSERT((g_cpu_lockset & (1 << cpu)) != 0);
         }
```
Comment (TaiJuWu) [inline sched/sched/sched_lock.c:145]: Look like scheduler lock can be token by two cpus?

Comment (hujun260) [reply sched/sched/sched_lock.c:145]: > Look like scheduler lock can be token by two cpus?

yes, The current implementation of sched_lock is that whenever any CPU acquires the sched_lock, it causes the sched_lock to be in effect for all CPUs. There will be subsequent patchs to address this issue.

Comment (TaiJuWu) [reply sched/sched/sched_lock.c:145]: Not related performance issue.
If there is any resource protected by `g_cpu_schedlock` but we can get scheduler lock with different cpus simultaneously we will face race issue.
If I am wrong, please correct me.

Comment (hujun260) [reply sched/sched/sched_lock.c:145]: This is a common misunderstanding about sched_lock. It cannot achieve mutual exclusion of resources, especially in SMP scenarios.
In most cases, if we want protect resources, enter_critical_section should be used.

**Review by TaiJuWu — APPROVED** [review]

File: include/nuttx/irq.h [hunk 3]
```diff
@@ -70,6 +70,18 @@
...
+#ifdef CONFIG_SMP
+#  define cpu_irqlock_clear() \
+  do \
+    { \
+      g_cpu_irqset = 0; \
+      SP_DMB(); \
+      g_cpu_irqlock = SP_UNLOCKED; \
+      SP_DSB(); \
+    } \
+  while (0)
```
Comment (TaiJuWu) [inline include/nuttx/irq.h:82]: nit: maybe it can be
```
#  define cpu_irqlock_clear(cpu) \
  do \
    { \
      g_cpu_irqset &= ~(1 << cpu); \
      SP_DMB(); \
      g_cpu_irqlock = SP_UNLOCKED; \
      SP_DSB(); \
    } \
  while (0)
```
similar with `cpu_irqlock_set(cpu)`
There is a benefit, it can be used to trace is there different cpus enter critical section simultaneously.
WDYT?

Comment (hujun260) [reply include/nuttx/irq.h:82]: "g_cpu_irqset = 0" can save at least one CPU instruction compared to "g_cpu_irqset &= ~(1 << cpu);"
If adding "DEBUGASSERT(g_cpu_irqset == 0);" in "cpu_irqlock_set(cpu)" can also achieve the effect you want?
for example:
#  define cpu_irqlock_set(cpu) \
  do \
    { \
      DEBUGASSERT(g_cpu_irqset == 0); \
      g_cpu_irqset |= (1 << cpu); \
    } \
  while (0)

Comment (TaiJuWu) [reply include/nuttx/irq.h:82]: Got it, thank for your explaination. This idea is very great. Thanks.
